File: tests/Web.AcceptanceTests/Pages/ProgressPage.cs
namespace LiftAndShift.Web.AcceptanceTests.Pages;

public class ProgressPage(IPage page) : BasePage(page)
{
    public override string PagePath => $"{BaseUrl}/progress";

    public Task AssertHeading() => Assertions.Expect(Page.Locator("h1")).ToHaveTextAsync("Progress");

    public async Task SelectExerciseByName(string name)
    {
        var select = Page.Locator("select").First;
        var value = await select.Locator("option", new() { HasText = name }).First.GetAttributeAsync("value");
        await select.SelectOptionAsync(new SelectOptionValue { Value = value });
    }

    public Task AssertTableRowVisible() => Assertions.Expect(Page.Locator("table tbody tr").First).ToBeVisibleAsync();

    public Task AssertNoDataMessageVisible() =>
        Assertions.Expect(Page.Locator("text=No completed sessions with this exercise yet.")).ToBeVisibleAsync();
}

File: src/Web/ClientApp/e2e/pages/progress.page.ts
import { expect } from '@playwright/test';
import { BasePage, baseUrl } from './base-page';

export class ProgressPage extends BasePage {
  get pagePath(): string {
    return `${baseUrl()}/progress`;
  }

  assertHeading() {
    return expect(this.page.locator('h1')).toHaveText('Progress');
  }

  async selectExerciseByName(name: string) {
    const select = this.page.locator('select').first();
    const value = await select.locator('option', { hasText: name }).first().getAttribute('value');
    await select.selectOption(value);
  }

  assertTableRowVisible() {
    return expect(this.page.locator('table tbody tr').first()).toBeVisible();
  }

  assertNoDataMessageVisible() {
    return expect(
      this.page.locator('text=No completed sessions with this exercise yet.')
    ).toBeVisible();
  }
}
